res/controllers/Denon-HC1000S-scripts.js

```diff
@@ -0,0 +1,869 @@
+                engine.setValue(group, "hotcue_" + cueIndex + "_position", positionInBeats);
+            }
+        } else {
+            engine.setValue(group, "hotcue_" + cueIndex + "_clear", 1);
+        }
+    }
+};
+
+DenonHC1000S.sampler = function(cueIndex, group, value, shift) {
+    if (!shift) {
+        if (DenonHC1000S.shiftSampleLeft && DenonHC1000S.shiftSampleRight) { //shiftSampleRight AND shiftSampleLeft buttons are pressed -> load currently selected track from library
+            engine.setValue("[Sampler" + cueIndex + "]", "LoadSelectedTrack", 1);
+        } else { // sample playback
+            print(group + ": Sampler#" + cueIndex + " selected; shiftLeft is " + DenonHC1000S.shiftSampleLeft + "; shiftRight is " + DenonHC1000S.shiftSampleRight);
+            engine.setValue("[Sampler" + cueIndex + "]", "cue_gotoandplay", 1);
+        }
+    } else { // shift pressed; stop or load sample
+
+        print("Shift pressed -> stopping " + group + ": Sampler#" + cueIndex + " shiftLeft is " + DenonHC1000S.shiftSampleLeft + "; shiftRight is " + DenonHC1000S.shiftSampleRight);
+        engine.setValue("[Sampler" + cueIndex + "]", "start_stop", 1);
+    }
+};
+DenonHC1000S.loopIn = function(group, value, shift) {
+    if (shift) {
+        engine.setValue(group, "loop_start_position", -1);
+    } else {
+        if (engine.getValue(group, "loop_enabled")) {
+            engine.setValue(group, "loop_in_goto", 1); // to fix LoopIn hanging
+        } else {
+            engine.setValue(group, "loop_in", 1);
+        }
+    }
+};
+
+DenonHC1000S.loopOut = function(midino, control, value, status, group) {
+    print("ButtonStatusLoopOut triggered");
+    var shift = DenonHC1000S.shift;
+    if (shift) {
+        print("loopOut: shift is pressed");
+        engine.setValue(group, "loop_out", -1);
+        engine.setValue(group, "loop_end_position", -1);
+    } else {
+        if (value === 64) { //only on press not on release to avoid doubles
+            engine.setValue(group, "loop_end_position", -1); //removes current to fix hanging out
+            engine.setValue(group, "loop_out", 1); //and sets again
+            print("value is " + value);
+        }
+    }
+};
+
+DenonHC1000S.reloop = function(group, value, shift) {
+    var loopInPosition = engine.getValue(group, "loop_start_position");
+    var loopOutPosition = engine.getValue(group, "loop_end_position");
+    if (loopInPosition !== -1 && loopOutPosition !== -1) {
+        engine.setValue(group, "reloop_toggle", 1);
+    } else {
+        var samplesPerBeat = DenonHC1000S.samplesPerBeat(group);
+        loopInPosition = engine.getValue(group, "playposition") * engine.getValue(group, "track_samples");
+        if (shift) {
+            var loopInPositionInBeats = loopInPosition / samplesPerBeat;
+            if ((loopInPositionInBeats - Math.floor(loopInPositionInBeats)) > 0.5) {
+                loopInPosition = Math.floor(0.5 + loopInPositionInBeats) * samplesPerBeat;
```

Comment:
You can replace all occurences of `Math.floor(0.5 + ...)` with `Math.round(...)`